Implement terraform-provider-tidbcloud (#4)
tidbcloud/terraform-provider-tidbcloud: README.md

> # Terraform TiDBCloud Provider

Reviewer: ```suggestion
# Terraform TiDB Cloud Provider
```

> [![License](https://img.shields.io/github/license/pingcap/tispark.svg)](https://github.com/pingcap/tispark/blob/master/LICENSE)

Reviewer: TiSpark link? not right. You should go to img.shields.io to generate a badge for this repo.

> This is the repository for the terraform-provider-tidbcloud, which allows one to use Terraform with TiDBCloud. Learn more about [TiDBCloud](https://en.pingcap.com/tidb-cloud/)

Reviewer: ```suggestion
This is the repository for the terraform-provider-tidbcloud, which allows one to use Terraform with TiDB Cloud. Learn more about [TiDB Cloud](https://en.pingcap.com/tidb-cloud/)
```

    * [Importing the restore cluster](#importing-the-restore-cluster)
    * [Destroy the dedicated cluster](#destroy-the-dedicated-cluster)
- [Developing the Provider](#developing-the-provider)
    * [Environment](#environment)
    * [Building the provider from source](#building-the-provider-from-source)
    * [Generate or update documentation in docs file](#generate-or-update-documentation-in-docs-file)
    * [Running the acceptance test](#running-the-acceptance-test)
    * [Debug the provider](#debug-the-provider)
- [Follow us](#follow-us)
    * [Twitter](#twitter)
- [License](#license)

## Requirements

- [Terraform](https://www.terraform.io/downloads.html) >= 1.0
- [Go](https://golang.org/doc/install) >= 1.18 (if you want to build the provider plugin)

## Support

Resources
- [cluster](./docs/resources/cluster.md)
- [backup](./docs/resources/backup.md) (not support update)
- [restore](./docs/resources/restore.md) (not support update and delete)

DataSource
- [project](./docs/data-sources/project.md)
- [cluster spec](./docs/data-sources/cluster_spec.md)
- [restore](./docs/data-sources/restore.md)
- [backcup](./docs/data-sources/backup.md)

## Using the provider

> Documentation about the provider use and the corresponding specific configuration options can be found on the [official's website](https://www.terraform.io/language/providers).

Reviewer: ```suggestion
Documentation about the provider usage and the corresponding specific configuration options can be found on the [official's website](https://www.terraform.io/language/providers).
```

> Here We just give an example to show how to use the TiDBCloud provider. 

Reviewer: ```suggestion
Here We just give an example to show how to use the TiDB Cloud provider. 
```

In this example, you will create and manage a dedicated cluster, create a backup for it and restore from the backup.

### Set up

> TiDBCloud provider has released to terraform registry. All you need to do is install terraform (>=1.0).

Reviewer: ```suggestion
TiDB Cloud provider has released to terraform registry. All you need to do is install terraform (>=1.0).
```

For Mac user, you can install it with Homebrew.

First, install the HashiCorp tap, a repository of all our Homebrew packages.
```shell
brew install hashicorp/tap/terraform
```
Now, install Terraform with hashicorp/tap/terraform.
```shell
brew install hashicorp/tap/terraform
```

> 

Reviewer: Two same cmds? Doesn't feel right. See https://www.terraform.io/downloads.

See [official doc](https://learn.hashicorp.com/tutorials/terraform/install-cli?in=terraform/aws-get-started) for other installation methods.

### Create an API key

The TiDB Cloud API uses HTTP Digest Authentication. It protects your private key from being sent over the network. 

However, terraform-provider-tidbcloud does not support managing API key now. So you need to create the API key in the [console](https://tidbcloud.com/console/clusters).

1. Click the account name in the upper-right corner of the TiDB Cloud console
2. Click Organization Settings. The organization settings page is displayed
3. Click the API Keys tab and then click Create API Key
4. Enter a description for your API key. The role of the API key is always Owner currently
5. Click Next. Copy and save the public key and the private key
6. Make sure that you have copied and saved the private key in a secure location. The private key only displays upon the creation. After leaving this page, you will not be able to get the full private key again

> 7. Click Done

Reviewer: Better just link https://docs.pingcap.com/tidbcloud/api/v1beta#section/Authentication/API-Key-Management instead of describing the steps ourselves.

> ### Get TiDBCloud provider

Reviewer: ```suggestion
### Get TiDB Cloud provider
```

> Create a main.tf file to get the TiDBCloud provider with API key:

Reviewer: ```suggestion
Create a main.tf file to get the TiDB Cloud provider with API key:
```

> Create a main.tf file to get the TiDB Cloud provider with API key:

Reviewer: I think see where "API key" occurs in this section.

```
terraform {
  required_providers {
    tidbcloud = {
      source = "tidbcloud/tidbcloud"
      version = "~> 1.0.0"
    }
  }
  required_version = ">= 1.0.0"
}
```

> - The `source` attribute defines the provider which will be downloaded from [Terraform Registry](https://registry.terraform.io/?_gl=1*r7mwys*_ga*NTgyMzMzOTU1LjE2NTk5NDA0MTA.*_ga_P7S46ZYEKW*MTY2MTQ4MTg5Mi4yNy4xLjE2NjE0ODIxODcuMC4wLjA.) by default

Reviewer: ```suggestion
- The `source` attribute defines the provider which will be downloaded from [Terraform Registry](https://registry.terraform.io/) by default
```

> - The `version` attribute is optional which defines the version of the provider, it will use the latest version in default
> - The `required_version` is optional which defines the version of the terraform, it will use the latest version in default

Reviewer: ```suggestion
- The `version` attribute is optional which defines the version of the provider, it will use the latest version by default
- The `required_version` is optional which defines the version of the terraform, it will use the latest version by default
```

### Config the provider

You need to config the provider like:

```
terraform {
  required_providers {
    tidbcloud = {
      source = "tidbcloud/tidbcloud"
      version = "~> 1.0.0"
    }
  }
  required_version = ">= 1.0.0"
}

provider "tidbcloud" {
  username = "fake_username"
  password = "fake_password"
}
```

username and password are the API key's public key and private key, you can also pass them with the environment:

```
export TiDBCLOUD_USERNAME = ${public_key}

> export TiDBCLOUD_PASSWORD = ${private_key}

Reviewer: Maybe all uppercase.

        }
      + id             = (known after apply)
      + name           = "firstCluster"
      + project_id     = "1372813089189561287"
      + region         = "eu-central-1"
      + status         = (known after apply)
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: 
```

Terraform will generate an execution plan for you: 
- You can check the diff between the configuration and the state
- You can also see the results of this `apply`: it will add a new resource, and no resource will be changed or destroyed
- The `known after apply` shows that you will get the value after `apply`

If everything is in your plan, type the `yes` to continue:

```
Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

tidbcloud_cluster.example_cluster: Creating...
tidbcloud_cluster.example_cluster: Creation complete after 1s [id=1379661944630234067]

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.

```

> Use `terraform show` or `terraform state show tidbcloud_cluster.example_cluste` to inspect the state of your resource. The former will show all the states. (all the resources and the data source)

Reviewer: ```suggestion
Use `terraform show` or `terraform state show tidbcloud_cluster.example_clustet` to inspect the state of your resource. The former will show all the states(all the resources and the data source).
```

```shell

> $ terraform state show tidbcloud_cluster.example_cluste

Reviewer: ```suggestion
$ terraform state show tidbcloud_cluster.example_cluster
```

    paused = false
    root_password = "Your_root_password1."
    port          = 4000
    ...
  }
```

After apply you will find the status turns to `RESUMING`:

```
# tidbcloud_cluster.example_cluster:
resource "tidbcloud_cluster" "example_cluster" {
    cloud_provider = "AWS"
    cluster_type   = "DEDICATED"
    config         = {
        components     = {
            tidb    = {
                node_quantity = 2
                node_size     = "8C16G"
            }
            tiflash = {
                node_quantity    = 2
                node_size        = "8C64G"
                storage_size_gib = 500
            }
            tikv    = {
                node_quantity    = 6
                node_size        = "8C32G"
                storage_size_gib = 500
            }
        }
        ip_access_list = [
            # (1 unchanged element hidden)
        ]
        paused         = false
        port           = 4000
        root_password  = "Your_root_password1."
    }
    id             = "1379661944630234067"
    name           = "firstCluster"
    project_id     = "1372813089189561287"
    region         = "eu-central-1"
    status         = "RESUMING"
}
```

> Wait for a moment, the status will be changed to AVAILABLE again.

Reviewer: ```suggestion
Wait for a moment, the status will be changed to `AVAILABLE` again.
```

                node_size        = "8C64G"
                storage_size_gib = 500
            }
            tikv    = {
                node_quantity    = 6
                node_size        = "8C32G"
                storage_size_gib = 500
            }
        }
        port          = 4000
        root_password = "Your_root_password1."
    }
    create_timestamp = "2022-08-26T08:16:33Z"
    id               = "780114"
    name             = "restoreCluster"
    project_id       = "1372813089189561287"
    status           = "PENDING"
}
```

You can see the restore task's status is `PENDING` and the cluster's status is `INITIALIZING`.

After the cluster is `AVAILABLE`, the restore task will be `RUNNING` and turn to `SUCCESS` at last.

It is everything ok? No, the bad news is the restored cluster is not managed by terraform. 

Don't worry, we can solve it in the next section.

### Importing the restore cluster

We can manage a cluster with terraform by import even if it is not created by terraform.

Let us import the cluster which is created by the restore task in the last section.

First add a cluster resource like:
```
resource "tidbcloud_cluster" "restore_cluster1" {}
```

> Then import the cluster by ` terraform import tidbcloud_cluster.restore_cluster1 projectId,clusterId`, you can get the projectId and clusterId by restore resource:

Reviewer: ```suggestion
Then import the cluster by `terraform import tidbcloud_cluster.restore_cluster1 projectId,clusterId`, you can get the projectId and clusterId by restore resource:
```

        port       = 4000
    }
    id             = "1379661944630264072"
    name           = "restoreCluster"
    project_id     = "1372813089189561287"
    region         = "eu-central-1"
    status         = "AVAILABLE"
}
```

In order to manage it, you can copy it to your config file. Remember to delete the status and id, for they are computed by terraform and can not be set in the config:
```
resource "tidbcloud_cluster" "restore_cluster1" {
      cloud_provider = "AWS"
      cluster_type   = "DEDICATED"
      config         = {
          components = {
              tidb    = {
                  node_quantity = 2
                  node_size     = "8C16G"
              }
              tiflash = {
                  node_quantity    = 2
                  node_size        = "8C64G"
                  storage_size_gib = 500
              }
              tikv    = {
                  node_quantity    = 6
                  node_size        = "8C32G"
                  storage_size_gib = 500
              }
          }
          port       = 4000
      }
      name           = "restoreCluster"
      project_id     = "1372813089189561287"
      region         = "eu-central-1"
}
```

> You can use `terraform fmt` to formate your config file:

Reviewer: ```suggestion
You can use `terraform fmt` to format your config file:
```

tidbcloud_cluster.example_cluster: Destroying... [id=1379661944630234067]
tidbcloud_cluster.example_cluster: Destruction complete after 0s
╷
│ Warning: Unsupported
│ 
│ restore can't be delete
╵

Destroy complete! Resources: 4 destroyed.
```

Note that a warning is appeared for restore can't be deleted.

If you execute `terraform show`, you will find nothing for all the states is cleared:
```
$ terraform show

```

## Developing the Provider

### Environment

If you wish to work on the provider, you'll first need [Go](http://www.golang.org) installed on your machine (see [Requirements](#requirements) above).

### Building the provider from source

1. Clone the repository
```shell
git clone git@github.com:tidbcloud/terraform-provider-tidbcloud
```
2. Enter the repository directory
```shell
cd terraform-provider-tidbcloud
```
3. Build the provider using the Go `install` command. This will build the provider and put the provider binary in the `$GOPATH/bin` directory.
```shell
go install
```

### Generate or update documentation in docs file

run `go generate`

### Running the acceptance test

see [here](./internal/README.md) for more detail.

### Debug the provider

> I will introduce how to debug with Terraform CLI development overrides. About other ways to debug the provider. See [terraform official doc](https://www.terraform.io/plugin/debugging) for more detail

Reviewer: ```suggestion
I will introduce how to debug with Terraform CLI development overrides. About other ways to debug the provider, see [terraform official doc](https://www.terraform.io/plugin/debugging) for more detail
```

Development overrides is a method of using a specified local filesystem Terraform provider binary with Terraform CLI, such as one locally built with updated code, rather than a released binary.

1. create `.terraformrc` in your operating system user directory
```
provider_installation {

  dev_overrides {
      "hashicorp.com/edu/tidbcloud" = "/usr/local/go/bin"
  }

  # For all other providers, install them directly from their origin provider
  # registries as normal. If you omit this, Terraform will _only_ use
  # the dev_overrides block, and so no other providers will be available.
  direct {}
}
```

2. run go install in the terraform-provider-tidbcloud, you will find the `terraform-provider-tidbcloud` will be installed under the `/usr/local/go/bin` 
```
go install
```

3. Terraform CLI commands, such as terraform apply, will now use the specified provider binary if you follow the below config:
```
terraform {
  required_providers {
    tidbcloud = {

>       source = "hashicorp/tidbcloud"

Reviewer: hashicorp/tidbcloud?

Author: we need not to use terraform-provider-tidbcloud, tidbcloud is enough. This is stated in the terraform documentation， but I can not find it now

    }
  }
}
```

## Follow us

Twitter [@PingCAP](https://twitter.com/PingCAP)

## License

> TiSpark is under the Apache 2.0 license. See the [LICENSE](./LICENSE) file for details.

Reviewer: TiSpark
